host/xtest/acipher_perf.c

```diff
@@ -0,0 +1,942 @@
+		prepare_keypair(width_bits, buf, blen, mode, algo, main_algo);
+
+	if (mode == MODE_SIGN || mode == MODE_VERIFY)
+		prepare_hash(size, main_algo);
+
+	if (main_algo == RSA && salt_len > 0) {
+		params[0].attributeID = TEE_ATTR_RSA_PSS_SALT_LENGTH;
+		params[0].content.value.a = salt_len;
+		params[0].content.value.b = 0;
+		param_count = 1;
+		res = pack_attrs(params, param_count, &buf, &blen);
+		CHECK(res, "pack_attrs", return -1;);
+	}
+
+	if (mode == MODE_GENKEYPAIR) {
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_CIPHER_PROCESS_GEN_KEYPAIR;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+						 TEEC_MEMREF_TEMP_INOUT,
+						 TEEC_NONE, TEEC_NONE);
+		op.params[0].value.a = width_bits;
+		op.params[0].value.b = l;
+		op.params[1].tmpref.buffer = buf;
+		op.params[1].tmpref.size = blen;
+	} else {
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_CIPHER_PROCESS;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_INPUT,
+						 TEEC_VALUE_INPUT,
+						 TEEC_MEMREF_PARTIAL_INPUT,
+						 TEEC_MEMREF_PARTIAL_OUTPUT);
+
+		op.params[0].tmpref.buffer = buf;
+		op.params[0].tmpref.size = blen;
+		op.params[1].value.a = l;
+		op.params[1].value.b = mode;
+
+		switch (mode) {
+		case MODE_ENCRYPT:
+			op.params[2].memref.parent = &in_shm;
+			op.params[2].memref.size = size;
+			op.params[3].memref.parent = &out_shm;
+			op.params[3].memref.size = out_shm.size;
+			break;
+		case MODE_SIGN:
+			op.params[2].memref.parent = &hash_shm;
+			op.params[2].memref.size = hash_shm.size;
+			op.params[3].memref.parent = &out_shm;
+			op.params[3].memref.size = out_shm.size;
+			break;
+		case MODE_DECRYPT:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			op.params[3].memref.parent = &in_shm;
+			op.params[3].memref.size = size;
+			break;
+		case MODE_VERIFY:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[2].memref.parent = &hash_shm;
+			op.params[2].memref.size = hash_shm.size;
+			op.params[3].memref.parent = &out_shm;
+			op.params[3].memref.size = out_shm.size;
+			break;
+		default:
+			fprintf(stderr, "Unexpected mode value\n");
+			exit(1);
+		}
+	}
+
+	if (warmup)
+		do_warmup(warmup);
+
+	while (n-- > 0) {
+		if (mode == MODE_ENCRYPT && main_algo == RSA) {
+			/* Make sure the N > M */
+			((unsigned char *)(in_shm.buffer))[0] = 0x00;
+			/* Avoid the problem that the last encryption result is
+			shorter than the plaintext. */
+			op.params[3].memref.size = out_shm.size;
+		}
+
+		get_current_time(&t0);
+
+		res = TEEC_InvokeCommand(&sess, cmd, &op, &ret_origin);
+
+		check_res(res, "TEEC_InvokeCommand()", &ret_origin);
+
+		get_current_time(&t1);
+
+		time += 1.0f * timespec_diff_ns(&t0, &t1) / 1000.0f / 1000.0f / 1000.0f;
+
+		update_stats(&stats, timespec_diff_ns(&t0, &t1));
+		if (n % (n0 / 10) == 0)
+			vverbose("#");
+	}
+
+	vverbose("\n");
+	sd = stddev(&stats);
+	printf("Ops=%.3f kops\n", n * l / time / 1000.0f);
```

Comment:
is it operation per seconds ?
if yes, should we do total time / (n * l) instead ?